src/lib/socialimagesdatabase.cpp

```diff
@@ -0,0 +1,448 @@
+                                        "FROM images "\
+                                        "WHERE accountId = :accountId");
+
+    if (olderThan.isValid()) {
+        queryString.append(" AND createdTime < :createdTime");
+    }
+
+    QSqlQuery query = q->prepare(queryString);
+    query.bindValue(":accountId", accountId);
+
+    if (olderThan.isValid()) {
+        query.bindValue(":createdTime", olderThan.toTime_t());
+    }
+
+    if (!query.exec()) {
+        qWarning() << Q_FUNC_INFO << "Failed to query images:" << query.lastError().text();
+        return data;
+    }
+
+    while (query.next()) {
+        data.append(SocialImage::create(query.value(0).toInt(),                             // accountId
+                                        query.value(1).toString(),                          // imageUrl
+                                        query.value(2).toString(),                          // imageFile
+                                        QDateTime::fromTime_t(query.value(3).toUInt())));   // createdTime
+    }
+
+    return data;
+}
+
+bool operator==(const SocialImage::ConstPtr &image1,
+                const SocialImage::ConstPtr &image2)
+{
+    return image1->accountId() == image2->accountId()
+            && image1->imageUrl() == image2->imageUrl();
+}
+
+SocialImagesDatabase::SocialImagesDatabase()
+    : AbstractSocialCacheDatabase(*(new SocialImagesDatabasePrivate(this)))
+{
+}
+
+SocialImagesDatabase::~SocialImagesDatabase()
+{
+    wait();
+}
+
+void SocialImagesDatabase::purgeAccount(int accountId)
+{
+    Q_D(SocialImagesDatabase);
+    QMutexLocker locker(&d->mutex);
+
+    d->queue.purgeAccounts.append(accountId);
+}
+
+SocialImage::ConstPtr SocialImagesDatabase::image(const QString &imageUrl) const
+{
+    Q_D(const SocialImagesDatabase);
+
+    if (d->queue.insertImages.contains(imageUrl)) {
+        return d->queue.insertImages.value(imageUrl);
+    }
+
+    QSqlQuery query = prepare(
+                "SELECT accountId, "
+                "imageUrl, imageFile, createdTime "
+                "FROM images WHERE imageUrl = :imageUrl");
+    query.bindValue(":imageUrl", imageUrl);
+    if (!query.exec()) {
+        qWarning() << Q_FUNC_INFO << "Error reading from images table:" << query.lastError();
+        return SocialImage::Ptr();
+    }
+
+    // If we have the image, we have a result, otherwise we won't have the image
+    if (!query.next()) {
+        return SocialImage::Ptr();
+    }
+
+    return SocialImage::create(query.value(0).toInt(),                           // accountId
+                               query.value(1).toString(),                        // imageUrl
+                               query.value(2).toString(),                        // imageFile
+                               QDateTime::fromTime_t(query.value(3).toUInt()));  // createdTime
+}
+
+void SocialImagesDatabase::removeImage(const QString &imageUrl)
```

Comment:
I wonder whether this should require accountId also.  Eg, what if one account's post uses an image, but that same image (url) is only used in another account's post from more than 2 weeks ago, or whatever.  We shouldn't purge the image, in this case, because one account is still using it.  I guess this means more data duplication in theory, but shouldn't be much in practice (ie, composite primary key consisting of accountId+imageUrl)